Fixed operation id support

robbert229 opened the pull request:
Currently there is an issue with the grpc-gen-swagger where it does not respect the operation_id that can be given in the proto file.

This is an example of what it may look like to fix the issue.

robbert229 commented:
This resolves #740

robbert229 commented:
is this done by doing `make examples` in the root directory? I did attempt to do this but there was a significant amount of changes that was generated that was unrelated to this bug.

johanbrandhorst commented:
Yes, `make examples`, but you need to ensure you're using the same version of `protoc` and `protoc-gen-go` when running it. Consult the Travis config and Gopkg.toml.

robbert229 commented:
I regenerated the examples! Thanks for the help

johanbrandhorst commented:
Could you please rebase on master and regenerate again? Looks like there's a merge issue.

askurydzin commented on protoc-gen-swagger/genswagger/template.go:
```diff
@@ -724,6 +724,13 @@ func renderServices(services []*descriptor.Service, paths swaggerPathsObject, re
+						operationObject.OperationID = opts.OperationId
+
+						if bIdx > 0 {
+							operationObject.OperationID = fmt.Sprintf("%s%d", opts.OperationId, bIdx)
```
bIdx+1 as per https://github.com/grpc-ecosystem/grpc-gateway/blob/8b937447b5fea1a0a60782b3cc42d279694f41fd/protoc-gen-swagger/genswagger/template.go#L703

robbert229 commented:
@johanbrandhorst I regenerated and rebased on master.

@askurydzin done. 

On another note, with the current structure of the program it is kind of hard to easily write a unit test for this. Normally if I found a bug and fixed it I would typically include a test to prevent regression. Would it be okay to potentially in another pr do a bit of refactoring (pull out some of the naming code into a separate function) and then write a test ensuring proper operation id calculation?

robbert229 commented:
ah. I rebased onto the master on my fork and not on the upstream. Now I have actually rebased...

johanbrandhorst commented:
Looks like there's a diff in some generated files: https://travis-ci.org/grpc-ecosystem/grpc-gateway/jobs/423385162. Could you please try again with the `protoc-gen-go` version locked in `Gopkg.toml`?

johanbrandhorst commented:
I've restarted the jobs

johanbrandhorst commented:
Sigh, not sure why this is so flaky recently

johanbrandhorst commented:
The last error looks like another generation error, could you take another look?

vtolstov commented:
@robbert229 can you take a look?

robbert229 commented:
Hmm. I am trying to figure out why it is failing. The build job for go 1.9 works fine, but for only does go 1.10 fail, and its because a different diff is generated. I don't have the expertise with Bazel, to understand what is going on here. Does anyone have any insight on what I might have done incorrectly?

```diff
@@ -452,9 +454,9 @@func _ABitOfEverything_OneofSizer(msg proto.Message) (n int) {
 // Nested is nested type.
 type ABitOfEverything_Nested struct {
 	// name is nested field.
-	Name                 string                           `protobuf:"bytes,1,opt,name=name" json:"name,omitempty"`
-	Amount               uint32                           `protobuf:"varint,2,opt,name=amount" json:"amount,omitempty"`
-	Ok                   ABitOfEverything_Nested_DeepEnum `protobuf:"varint,3,opt,name=ok,enum=grpc.gateway.examples.examplepb.ABitOfEverything_Nested_DeepEnum" json:"ok,omitempty"`
+Name                 string                           `protobuf:"bytes,1,opt,name=name,proto3" json:"name,omitempty"`
+Amount               uint32                           `protobuf:"varint,2,opt,name=amount,proto3" json:"amount,omitempty"`
+Ok                   ABitOfEverything_Nested_DeepEnum `protobuf:"varint,3,opt,name=ok,proto3,enum=grpc.gateway.examples.examplepb.ABitOfEverything_Nested_DeepEnum" json:"ok,omitempty"`
 	XXX_NoUnkeyedLiteral struct{}                         `json:"-"`
 	XXX_unrecognized     []byte                           `json:"-"`
 	XXX_sizecache        int32                            `json:"-"`
```

johanbrandhorst commented:
I think its a `protoc-gen-go` or `protoc` version mismatch. Make sure you've got the same versions used by .Travis.yml

vtolstov commented:
gentle ping again...

johanbrandhorst commented:
Please rebase on master for new CI functionality.

mestrade commented:
Sorry if my message is not at the good place, but I've done a quick fix to make the operationID generated by default with the service name instead of only the call name. Fix look like

`if bIdx == 0 {
					operationObject.OperationID = fmt.Sprintf("%s%s", svc.GetName(), meth.GetName())
				} else {
					// OperationID must be unique in an OpenAPI v2 definition.
					operationObject.OperationID = fmt.Sprintf("%s%s%d", svc.GetName(), meth.GetName(), bIdx+1)
				}
`

Is this something you could consider ? if yes I can do a PR

johanbrandhorst commented:
@mestrade Why don't you open a PR and we can discuss it in there :).

mestrade commented:
Sorry to ask this, but it seems I cannot push my branch to create the PR, do I need some special rights ? my dev env. is having some special git setup so it's maybe also broken here :/

johanbrandhorst commented:
No worries @mestrade, I think you may benefit from reading the github contribution guidelines: https://help.github.com/articles/creating-a-pull-request-from-a-fork/

johanbrandhorst commented:
You need to create a fork first, you cannot commit directly to this repo.
